addon/internal-session.js

```diff
@@ -44,26 +44,34 @@ export default Ember.ObjectProxy.extend(Ember.Evented, {
 
   restore() {
     return new Ember.RSVP.Promise((resolve, reject) => {
-      let restoredContent   = this.store.restore();
-      let { authenticator } = (restoredContent.authenticated || {});
-      if (!!authenticator) {
-        delete restoredContent.authenticated.authenticator;
-        this.container.lookup(authenticator).restore(restoredContent.authenticated).then((content) => {
-          this.set('content', restoredContent);
-          this._setup(authenticator, content);
-          resolve();
-        }, () => {
-          Ember.Logger.debug(`The authenticator "${authenticator}" rejected to restore the session - invalidating…`);
+      let restored = this.store.restore();
+      if (typeof restored === 'undefined' || typeof restored.then === 'undefined') {
+        restored = Ember.RSVP.Promise.resolve(restored || {});
+      }
+      restored.then((restoredContent) => {
+        let { authenticator } = (restoredContent.authenticated || {});
+        if (!!authenticator) {
+          delete restoredContent.authenticated.authenticator;
+          this.container.lookup(authenticator).restore(restoredContent.authenticated).then((content) => {
+            this.set('content', restoredContent);
+            this._setup(authenticator, content);
+            resolve();
+          }, () => {
+            Ember.Logger.debug(`The authenticator "${authenticator}" rejected to restore the session - invalidating…`);
+            this.set('content', restoredContent);
+            this._clear();
+            reject();
+          });
+        } else {
+          delete (restoredContent || {}).authenticated;
           this.set('content', restoredContent);
           this._clear();
           reject();
-        });
-      } else {
-        delete (restoredContent || {}).authenticated;
-        this.set('content', restoredContent);
+        }
+      }, () => {
         this._clear();
         reject();
-      }
+      });
     });
   },
 
```

Comment:
`_setup` and `_clear` are now async too. `setUnknownProperty` should technically be async too, but Ember doesn't expect it to return a promise so it will persist property changes asynchronously but the methods will return directly.